### PR TITLE
[Spec Decode] Make DFlash draft compute honor dynamic K

### PR DESCRIPTION
**Pull request by RichApple123**
## Purpose

Make DFlash's physical draft workload honor the batch-uniform speculative K
selected by the existing dynamic speculative-decoding scheduler.

Today the scheduler can request fewer than the configured maximum draft
tokens, but DFlash still builds and executes its maximum query shape before
the worker discards unused proposals. This change:

- carries the scheduled K in `ExecuteModelState`, binding it to the matching
  execute/sample iteration under async scheduling;
- skips proposal generation for K=0;
- advertises dynamic physical shapes as a speculator capability, enabled only
  for DFlash;
- sizes DFlash query inputs, attention metadata, sampling buffers, and CUDA
  graph dispatch from the scheduled K; and
- publishes exactly K proposals while retaining the configured global maximum
  for allocation and graph coverage.

Unsupported speculators keep their existing maximum physical shape. DSpark
explicitly opts out because its sequential head needs separate model-specific
validation. The common K=0 skip applies to all speculators.

### Related work

This does not duplicate #48692. That PR implements per-request adaptive
verification and explicitly initializes DFlash with
`allow_dynamic_decode_shapes=False`; this PR shortens the batch-uniform
DFlash drafter computation selected by the existing scheduler. It is also
separate from the per-request proposal API in #48202/#48204.

PR #47737 handles DSpark K=0 graph capture, #46399 disables dynamic schedules
for fixed-K proposers, and #48329 addresses autoregressive graph selection.
This PR does not replace those changes.

## Test Plan

```bash
pytest -q \
  tests/v1/worker/test_gpu_model_runner_v2_spec_decode.py \
  tests/v1/worker/test_gpu_model_runner_v2_eplb.py

mypy tests/v1/worker/test_gpu_model_runner_v2_spec_decode.py
```

Relevant Ruff, Ruff format, and typos pre-commit hooks were run on every
changed file.

GPU validation uses one NVIDIA H100 80GB, Qwen3.5-4B with
Qwen3.5-4B-DFlash, Model Runner V2, async scheduling, compiled CUDA graphs,
and K=0/1/3/7. The serving matrix covers concurrency 1/64/128/256 with three
repeats per point.

## Test Result

- Worker tests: 7 passed.
- Mypy: no issues.
- Relevant pre-commit hooks: passed.
- Serving smoke: 48/48 runs completed, zero failed requests, and DFlash full
  CUDA graph capture/replay succeeded for K=1/3/7; K=0 skipped the drafter.

A five-repeat natural mixed-prompt A/B isolates the physical-shape change:

| Batch | K=0 TPS delta | K=1 TPS delta | K=3 TPS delta | K=7 control |
|---:|---:|---:|---:|---:|
| 64 | +37.3% | +15.3% | +2.3% | -3.7% |
| 128 | +40.0% | +5.1% | +4.4% | -2.6% |
| 256 | +36.9% | +6.2% | +4.6% | +0.4% |

K=7 retains the original maximum physical shape and is the no-op control; its
small positive/negative deltas indicate run/system variance. The lower-K gains
come from removing unused drafter work. Acceptance length for each fixed K is
stable across batch sizes (K=1: 1.879-1.882, K=3: 3.158-3.160, K=7:
4.415-4.429).

An independent standard Sonnet serving sweep covers K=0/1/3/7 and concurrency
1/64/128/256 with three repeats per point (48 runs, zero failures). At
concurrency 1, robust TPS rises from 204.33 at K=0 to 317.27 at K=7 (+55.3%).
At concurrency 64, K=0 reaches 3511.24 while the best positive candidate,
K=3, reaches 3042.50. TPS, `tpot:50` goodput, and mean TPOT therefore all
select the genuinely dynamic schedule `[[1, 63, 7], [64, 256, 0]]`.

The global maximum K still controls worker allocation, scheduler lookahead,
and KV-cache reservation. This patch reduces draft compute, not reserved KV
memory, and does not add an online K-selection policy.

## AI assistance

This patch was developed with OpenAI Codex assistance and the commit includes
the required attribution trailer. This is a Draft PR; the human submitter must
review every changed line and confirm the test/benchmark evidence before
marking it ready for review.

### CODE REVIEW

**Comment by benchislett** [comment]
This is dangerous as DFlash non-causal attention can yield lower acceptance rates when running with too few draft tokens.

**Comment by RichApple123** [comment]
Closing this draft after re-evaluating the algorithmic assumption. For non-causal DFlash, shortening the physical draft block is not equivalent to computing the full trained block and truncating verification: later mask positions can influence earlier draft logits and acceptance. The maintainer concern is valid, and this should not be presented as removal of unused compute. I am preserving the branch and benchmark artifacts for reference, but will not continue this design as a general DFlash optimization. Thank you for catching this.
